dask/highlevelgraph.py

```diff
@@ -301,29 +314,84 @@ def __dask_distributed_pack__(self, client) -> Optional[Any]:
 
-        Alternatively, the method can return None, which will make Distributed
-        materialize the layer and use a default packing method.
+        The default implementation materialize the layer thus layers such as Blockwise
+        and ShuffleLayer should implement a specialized pack and unpack function in
+        order to avoid materialization.
 
         Parameters
         ----------
         client: distributed.Client
             The client calling this function.
+        client_keys : Iterable
+            List of keys requested by the client.
 
         Returns
         -------
         state: Object serializable by msgpack
-            Scheduler compatible state of the layer or None
+            Scheduler compatible state of the layer
         """
-        return None
+        from distributed.client import Future
+        from distributed.utils_comm import unpack_remotedata, subs_multiple
+        from distributed.worker import dumps_task
+        from distributed.utils import CancelledError
+
+        dsk = dict(self)
+
+        # Find aliases not in `client_keys` and substitute all matching keys
+        # with its Future
+        values = {
+            k: v
+            for k, v in dsk.items()
+            if isinstance(v, Future) and k not in client_keys
+        }
+        if values:
+            dsk = subs_multiple(dsk, values)
+
+        # Unpack remote data and record its dependencies
+        dsk = {k: unpack_remotedata(v, byte_keys=True) for k, v in dsk.items()}
+        unpacked_futures = set.union(*[v[1] for v in dsk.values()]) if dsk else set()
+        for future in unpacked_futures:
+            if future.client is not client:
+                raise ValueError(
+                    "Inputs contain futures that were created by another client."
+                )
+            if stringify(future.key) not in client.futures:
+                raise CancelledError(stringify(future.key))
+        unpacked_futures_deps = {}
+        for k, v in dsk.items():
+            if len(v[1]):
+                unpacked_futures_deps[k] = {f.key for f in v[1]}
+        dsk = {k: v[0] for k, v in dsk.items()}
+
+        # Calculate dependencies without re-calculating already known dependencies
+        missing_keys = set(dsk.keys()).difference(known_key_dependencies.keys())
+        dependencies = {
+            k: keys_in_tasks(all_hlg_keys, [dsk[k]], as_list=False)
+            for k in missing_keys
+        }
+        for k, v in unpacked_futures_deps.items():
+            dependencies[k] = set(dependencies.get(k, ())) | v
+
+        # The scheduler expect all keys to be strings
+        dependencies = {
+            stringify(k): {stringify(dep) for dep in deps}
+            for k, deps in dependencies.items()
+        }
+
+        all_hlg_keys = all_hlg_keys.union(dsk)
+        dsk = {
+            stringify(k): stringify(v, exclusive=all_hlg_keys) for k, v in dsk.items()
+        }
```

Comment:
all_hlg_keys is of type Iterable, which doesn't have a .union method. AbstractSet in the function parameters + a change in variable name here is necessary.

```suggestion
        merged_hlg_keys = all_hlg_keys | dsk.keys()
        dsk = {
            stringify(k): stringify(v, exclusive=merged_hlg_keys) for k, v in dsk.items()
        }
```